### PR TITLE
feat(core): add support for TypeScript 5.0

### PR DESCRIPTION
**Pull request by crisbeto**
Updates the project to support TypeScript 5.0 and to resolve any errors that came up as a result of the update.

### CODE REVIEW

**Comment by crisbeto** [comment]
[Passing TGP](http://test/OCL:510374747:BASE:510401277:1676638241865:b3188925)

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_simple_partial.js [hunk 1]
```diff
@@ -5,7 +5,7 @@
 // TODO: Work out how to fix the broken segment for the last item in a template
 .ɵɵelem // SOURCE: "/ng_if_simple.ts" "</div>'"
 …
-.ɵɵtextInterpolate(ctx_r0.name);\n  }\n}\n\n // SOURCE: "/ng_if_simple.ts" "{{ name }}"
+.ɵɵtextInterpolate(ctx_r0.name);\n  }\n}\n\nclass  // SOURCE: "/ng_if_simple.ts" "{{ name }}"
```
Comment (devversion) [inline packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_simple_partial.js:8]: Does this mean that the source mappings are incorrect? well, I guess, they were a little incorrect anyway if they included the `\n}\n`?

File: packages/service-worker/worker/testing/events.ts [hunk 2]
```diff
@@ -7,10 +7,10 @@
...
 
 export class MockEvent implements Event {
-  readonly AT_TARGET = -1;
-  readonly BUBBLING_PHASE = -1;
-  readonly CAPTURING_PHASE = -1;
-  readonly NONE = -1;
+  readonly AT_TARGET = 2;
+  readonly BUBBLING_PHASE = 3;
+  readonly CAPTURING_PHASE = 1;
+  readonly NONE = 0;
```
Comment (devversion) [inline packages/service-worker/worker/testing/events.ts:13]: Is there any context on this change? (if you recall, can be in the comment here)

File: packages/compiler-cli/package.json [hunk 3]
```diff
@@ -53,7 +53,7 @@
   },
   "peerDependencies": {
     "@angular/compiler": "0.0.0-PLACEHOLDER",
-    "typescript": ">=4.8.2 <5.0"
+    "typescript": ">=4.8.2 <5.1"
```
Comment (devversion) [inline packages/compiler-cli/package.json:56]: Just picking this place for the comment: Are there things we need to keep in mind with the decorator changes. i.e. when applications don't have `--experimentalDecorators=true`? https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#differences-with-experimental-legacy-decorators

**Review by devversion — APPROVED** [review]
LGTM, just a few small questions on additional context.

File: packages/service-worker/worker/testing/events.ts (hunk 2 above)
Comment (crisbeto) [reply packages/service-worker/worker/testing/events.ts:13]: The internal typings changed from `readonly NONE: number` to `readonly NONE: 0` so our mock had to be updated as well.

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_simple_partial.js (hunk 1 above)
Comment (crisbeto) [reply packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_simple_partial.js:8]: tbh, I'm not sure, as you said a correct mapping wouldn't include the newlines. It's possible that this is just how our testing utilities work.

File: packages/compiler-cli/package.json (hunk 3 above)
Comment (crisbeto) [reply packages/compiler-cli/package.json:56]: For the moment if apps disable `experimentalDecorators`, they'll get a compilation error on parameter decorators before the code even reaches our compiler. I think that eventually we'll have a schematic to move apps away from constructor injection.

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 99d874fe3b486f3669b0e8f1910e31c4fa278308.

**Pull request closed without merging**